Add V-Sync & Borderless options to video settings

Himanshu-01 opened the pull request:
![image](https://github.com/user-attachments/assets/64b2d272-b263-4032-bab6-beb6e6783b86)
![image](https://github.com/user-attachments/assets/1a779de1-0619-490a-a389-d27778eccb65)
![image](https://github.com/user-attachments/assets/9daca029-78c1-4bdb-b3e7-1c27dd6b3eab)


* Allows players to tweak vertical-sync from game itself
* Disables the shell_pcc_vsync from toggling vsync setting
* Implements new `c_screen_vsync_menu` custom ui

Himanshu-01 committed:
Add `s_shell_pcc_settings`

move out pcc functions to seperate header

Himanshu-01 committed:
Implement and use `c_screen_video_settings`

Himanshu-01 committed:
Add `V-Sync` option to video settings ui

Implement `c_screen_vsync_menu` custom class and disable shell_pcc_vsync setting for now

Himanshu-01 committed:
Implement and use `c_screen_multiplayer_video_settings`

remove original usage of `s_list_item_datum` and construct using static datum elements

Himanshu-01 committed:
Add `V-Sync` option to in-game video settings ui

Himanshu-01 committed:
Make vsync menu item focus to reflect present value

Berthalamew commented on xlive/Blam/Engine/interface/screens/screen_multiplayer_video_settings_menu.cpp:
```diff
@@ -0,0 +1,231 @@
+
+/* macro defines */
+
+#define k_mp_video_setting_list_name "mp video settings game list"
```
Should be a static const char* or static const char [] if this string isn't being appended to anything else at compile time

Berthalamew commented on xlive/Blam/Engine/interface/screens/screen_vsync_setting.cpp:
```diff
@@ -0,0 +1,170 @@
+	L"垂直同期",
+	L"V-Sync",
+	L"Synchronisation V",
+	L"sincronización vertical",
```
Should start with a capital letter?

Berthalamew commented on xlive/Blam/Engine/interface/screens/screen_video_settings.h:
```diff
@@ -1,8 +1,58 @@
+
+/* macro defines */
+
+#define k_no_of_visible_items_for_video_settings 8+1
```
should be a private enum

nukeulater commented on xlive/Blam/Engine/interface/screens/screen_video_settings.h:
```diff
@@ -1,8 +1,58 @@
+
+/* macro defines */
+
+#define k_no_of_visible_items_for_video_settings 8+1
```
always use parentheses in macros

Kantanomo replied:
I learned this the hard way.

Himanshu-01 committed:
address requested changes

move V-sync item below Resolution item

Himanshu-01 committed:
Add `struct s_custom_item_text_mapping`

add back original usage of `s_list_item_datum` in c_multiplayer_video_settings_list that uses custom text mapping for adding vsync string

nukeulater commented on xlive/Blam/Engine/interface/screens/screen_vsync_setting.cpp:
```diff
@@ -0,0 +1,170 @@
+
+/* public code */
+
+c_vsync_edit_list::c_vsync_edit_list(int16 user_flags) :
```
flags should always be of unsigned type

Berthalamew committed:
adjust s_custom_item_text_mapping

- allows us to create constant variables of s_custom_item_text_mapping
- make the  video setting item map a global constant

Himanshu-01 committed:
make user_flags unsigned

Himanshu-01 committed:
fix video settings not saving to registry

Himanshu-01 committed:
Add borderless mode to display settings

- add enum e_display_mode
- fix  c_user_interface_widget::setup_children() return type

Berthalamew commented on xlive/Blam/Engine/rasterizer/rasterizer_settings.h:
```diff
@@ -18,6 +18,15 @@ enum e_display_aspect_ratio : int32
 	_aspect_ratio_16x10 = 2
 };
 
+enum e_display_mode : int32
```
This is just a duplicate of e_rasterizer_window_mode, e_rasterizer_window_mode is the correctly named enum.
This should be removed.

Himanshu-01 committed:
fix borderless getting forced in startup

- remove e_display_mode

Himanshu-01 committed:
fix refresh rate set to 0 when changing display mode

Berthalamew committed:
fix window alignment when switching from windowed/fullscreen to borderless

Himanshu-01 committed:
disable refresh rate selection menu in borderless mode

Himanshu-01 committed:
add new borderless window patch

- fix various fullscreen bugs
- add _rasterizer_window_mode_borderless

Himanshu-01 committed:
remove old borderless code and hotkeys

Berthalamew commented on xlive/Blam/Engine/interface/screens/screen_display_mode.cpp:
```diff
@@ -48,6 +48,48 @@ enum e_display_mode_settings_list_items : uint16
 
 /* private code */
 
+static e_rasterizer_window_mode item_to_display_mode(e_display_mode_settings_list_items item)
```
Private code should be below public code.
Add prototypes for these to the prototype section.

Himanshu-01 committed:
disable/hide vsync option when not in fullscreen window

also add user_interface_utilities_play_sound

Berthalamew committed:
fix compiling when including user_interface_shared_globals.h`

- duplicate structs and constants existed between it and user_interface_screen_widget_definition.h
